### DISCUSSION TITLE
Can someone explain what these lines mean?

### DISCUSSION THREAD

**Discussion by joewinbpaul** [cssbattle/forum#371] [Q&A] [answered]
I was scrolling through codepen once and I saw this solution to Target #83 by Ilya Streltsyn. So I copied it and made a few changes to reduce characters (and got in the Top 10), but I don't understand most of the code, (which makes me feel like I don't deserve the Top 10th position).

![Screenshot (113)](https://github.com/cssbattle/forum/assets/102139966/74c01016-a83d-4649-b604-3d9cbbfd779b)

Can someone explain the code?

**Comment by SelenIT** [comment, marked as answer]
That demo was intended to be part of [the blog post](https://css-live.ru/tricks/border-image-gradients-sorcery.html) (in Russian, but you can use Google translate) showing how the `border-image` property can create such partially overlapping shapes. Here is [a fork of it with English comments](https://codepen.io/SelenIT/pen/abXeqye) (hope it will explain it a bit better).

The `border-image` shorthand consists of five sub-properties:
- `border-image-source` - the image itself, radial gradient in our case
- `border-image-slice` - sets the lines that slice the image into 4 to 9 slices that will be used as "tiles" to cover the area
- `border-image-width` - sets the widths of the outer "tiles", similar to the regular `border-width` (the value after the 1st `/`)
- `border-image-outset` - allows to expand the image beyond the border edge of the element (the value after the 2nd `/`, my favorite part of this property:)
- `border-image-repeat` - allows to repeat parts of the image as a pattern (not used in this example)

In our case, `border-image-outset: 53q 0` extends the image by ~50px above and below the border edges, making it a square (400x400 for `html` and 120x120 for `body`). Then `border-image-slice` cuts out squares of the 254x254px size (unitless numbers in `border-image-slice` mean "original pixels of the image", for gradients, they are the same as CSS pixels), and these squares are scaled down to the `border-image-width` actual value (which is clamped by 200px because two "tiles" have to fit into 400px total width anyway). You can see how it works in [this interactive animation](https://codepen.io/SelenIT/pen/mdBYMyK) (values above can be changed to see what they are affecting).

For `body`, `border-image-slice` is changed to 0 (via a CSS variable), meaning that nothing is cut from the sides and only the central "tile" made of the whole image is left. The keyword `fill` after the number in `border-image-slice` means that this "tile" is used (basically like another background), otherwise it would be ignored.

The last trick makes the same gradient declaration create either light blue or yellow circles depending on the font size. With default font size, 4em (64px) is larger than 32q, so the light blue part leaves no place for yellow. For `body`, on the other hand, `font-size` is set to 0, so 4em turn into 0 and the blue part is not rendered, leaving only the yellow one. This "adjustable gradient" trick can be helpful in many cases, not only in CSSBattle (e.g. for [cross-browser gradient animation](https://codepen.io/SelenIT/pen/bGLWYjO)).

Hope that this helped a bit:). Also please check out [the collection of other](https://codepen.io/collection/kNeZEV) interesting (imho) use cases of `border-image` :)
